File: backend/services/user_service.js
import { query } from "../utils/errors.js";

// Services accept a request-scoped database client and verified user ID.
// They do not depend on Express req/res objects, making them reusable.
export function getProfile(db, userId) {
  return query(
    db.from("profiles").select("*").eq("user_id", userId).maybeSingle(),
  );
}
export function saveProfile(db, userId, body) {
  return query(
    db
      .from("profiles")
      .upsert({
        ...body,
        user_id: userId,
        updated_at: new Date().toISOString(),
      })
      .select()
      .single(),
  );
}

File: backend/services/user_service.py
"""Profile persistence. Passwords and login email stay in Clerk."""

from datetime import datetime, timezone
from utils.errors import execute


def get_profile(db, user_id):
    rows = execute(db.table("profiles").select("*").eq("user_id", user_id).limit(1))
    return rows[0] if rows else None


def save_profile(db, user_id, body):
    rows = execute(
        db.table("profiles").upsert(
            {
                **body.model_dump(),
                "user_id": user_id,
                "updated_at": datetime.now(timezone.utc).isoformat(),
            }
        )
    )
    return rows[0]
